pkg/obfuscate/sql.go

```diff
@@ -16,9 +16,114 @@ import (
+	collectTableNames bool
+	collectCommands   bool
+	collectComments   bool
+	replaceDigits     bool
+
+	// size holds the byte size of the metadata collected by the filter.
+	size int64
+	// tablesSeen keeps track of unique table names encountered by the filter.
+	tablesSeen map[string]struct{}
+	// tablesCSV specifies a comma-separated list of tables.
+	tablesCSV strings.Builder
+	// commands keeps track of commands encountered by the filter.
+	commands []string
+	// comments keeps track of comments encountered by the filter.
+	comments []string
+}
+
+func (f *metadataFinderFilter) Filter(token, lastToken TokenKind, buffer []byte) (TokenKind, []byte, error) {
+	if f.collectComments && token == Comment {
+		// A comment with line-breaks will be brought to a single line.
+		comment := strings.TrimSpace(strings.Replace(string(buffer), "\n", " ", -1))
+		f.size += int64(len(comment))
+		f.comments = append(f.comments, comment)
+	}
+	if f.collectCommands {
+		switch token {
+		case Select, Update, Insert, Delete, Join, Alter, Drop, Create, Grant, Revoke, Commit, Begin, Truncate:
+			command := strings.ToUpper(token.String())
+			f.size += int64(len(command))
+			f.commands = append(f.commands, command)
+		}
+	}
+	if f.collectTableNames {
+		switch lastToken {
+		case From, Join:
+			// SELECT ... FROM [tableName]
+			// DELETE FROM [tableName]
+			// ... JOIN [tableName]
+			if r, _ := utf8.DecodeRune(buffer); !unicode.IsLetter(r) {
+				// first character in buffer is not a letter; we might have a nested
+				// query like SELECT * FROM (SELECT ...)
+				break
+			}
+			fallthrough
+		case Update, Into:
+			// UPDATE [tableName]
+			// INSERT INTO [tableName]
+			tableName := string(buffer)
+			if f.replaceDigits {
+				tableNameCopy := make([]byte, len(buffer))
+				copy(tableNameCopy, buffer)
+				tableName = string(replaceDigits(tableNameCopy))
+			}
+			f.storeTableName(tableName)
+			return TableName, buffer, nil
+		}
+	}
+	return token, buffer, nil
+}
+
+func (f *metadataFinderFilter) storeTableName(name string) {
+	if _, ok := f.tablesSeen[name]; ok {
+		return
+	}
+	if f.tablesSeen == nil {
+		f.tablesSeen = make(map[string]struct{}, 1)
+	}
+	f.tablesSeen[name] = struct{}{}
+	if f.tablesCSV.Len() > 0 {
+		f.tablesCSV.WriteByte(',')
+	}
+	f.tablesCSV.WriteString(name)
+}
+
+// Size returns the byte size of the metadata collected by the filter.
+func (f *metadataFinderFilter) Size() int64 {
+	return f.size + int64(len(f.tablesCSV.String()))
+}
+
+// Results returns metadata collected by the filter for an SQL statement.
+func (f *metadataFinderFilter) Results() SQLMetadata {
+	return SQLMetadata{
+		Size:      f.Size(),
+		TablesCSV: f.tablesCSV.String(),
+		Commands:  f.commands,
+		Comments:  f.comments,
+	}
+}
+
+// Reset implements tokenFilter.
+func (f *metadataFinderFilter) Reset() {
+	for k := range f.tablesSeen {
+		delete(f.tablesSeen, k)
+	}
+	f.size = 0
+	f.tablesCSV.Reset()
+	f.commands = nil
+	f.comments = nil
```

Comment:
```suggestion
	f.commands = f.commands[:0]
	f.comments = f.comments[:0]
```
This will keep the underlying array for the subsequent calls. Otherwise, with `nil`, we keep garbage collecting and re-allocating on every run, even though we could reuse what we started. WDYT?